Implement submit message in hub-rest client

piratekev opened the pull request:
This PR provides the ability to submit messages to hubs using the REST API

TODO: Add more examples (react, follow, verify)

piratekev committed:
feat(farcaster-js-hub-rest): implement submit message

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js

piratekev commented on .husky/pre-commit:
```diff
@@ -4,7 +4,10 @@
 
 # recreate generated files
-yarn workspaces foreach --all run generate
+yarn workspaces foreach --all run generate:openapi-axios
```
[Building deps in parallel causes an rpc ping timeout #5559
](https://github.com/nrwl/nx/issues/5559)

piratekev commented on packages/farcaster-js-hub-rest/src/openapi/post-process.sh:
```diff
@@ -0,0 +1,3 @@
+sed 's/File/Buffer/g' src/openapi/generated/apis/submit-message-api.ts > src/openapi/generated/apis/submit-message-api-temp.ts
```
The generated code for binary string is of type File, which is [not supported by version 3](https://github.com/OpenAPITools/openapi-generator/blob/master/docs/generators/typescript-axios.md#data-type-feature). This script replaces the occurrences of `File` with `Buffer` to make this work.

gsgalloway replied:
Nice. I didn't know about this post-processing hook

piratekev committed:
feat(farcaster-js-hub-rest): implement remaining submit message types

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js

piratekev committed:
chore(farcaster-js-hub-rest): remove only from tests

gsgalloway commented on packages/farcaster-js-hub-rest/src/hubRestApiClient.ts:
```diff
@@ -133,6 +155,301 @@ export class HubRestAPIClient {
+    link: {
+      type: string
+      displayTimestamp?: number
+      targetFid?: number
```
Weird that this can be undefined

piratekev replied:
Yeah I wasn't sure if I should expose it the way it works or just call is followUser and require it.

gsgalloway replied:
Mmm, yeah. Could go either way. Maybe a simple `followUser` method that just wraps `submitLink` gets us both worlds?

gsgalloway commented on packages/farcaster-js-hub-rest/src/hubRestApiClient.ts:
```diff
@@ -133,6 +155,301 @@ export class HubRestAPIClient {
+      type: 'like' | 'recast'
+      targetFid?: number
+      targetHash?: string
+      targetUrl?: string
```
I see that this is how the type is defined in @farcaster/core. I wonder though if we should hide this from the user and instead give them something like
```typescript
public async submitReaction(
  reaction: {
    type: 'like' | 'recast'
    target: CastID | { url: string }
  }
)
```
since these should be mutually exclusive IIRC

gsgalloway commented on packages/farcaster-js-hub-rest/src/hubRestApiClient.ts:
```diff
@@ -133,6 +155,301 @@ export class HubRestAPIClient {
+    link: {
+      type: string
+      displayTimestamp?: number
+      targetFid?: number
```
Also weird

gsgalloway commented on packages/farcaster-js-hub-rest/tests/integration/hubRestAPIClient.ts:
```diff
@@ -19,10 +19,17 @@ import {
 chai.use(chaiAsPromised);
 
 const userGaviFid = 69; // @gavi
+const userGaviBotFid = 6365; // @gavi
```
```suggestion
const userGaviBotFid = 6365; // @gavi-bot
```

gsgalloway commented on packages/farcaster-js-hub-rest/tests/integration/hubRestAPIClient.ts:
```diff
@@ -51,6 +59,103 @@ describe('HubWebClient', function() {
   });
 
+  if (signerPrivateKey !== undefined && signerPrivateKey !== '') {
+    // eslint-disable-next-line no-only-tests/no-only-tests
```
```suggestion
```

gsgalloway commented on packages/farcaster-js-neynar/docs/classes/v1.CastApi.md:
```diff
@@ -101,7 +101,7 @@ BaseAPI.configuration
 ### allCastsInThread
 
-▸ **allCastsInThread**(`requestParameters`, `options?`): `Promise`<`AxiosResponse`<[`AllCastsInThreadResponse`](../interfaces/v1.AllCastsInThreadResponse.md), `any`\>\>
+▸ **allCastsInThread**(`requestParameters`, `options?`): `Promise`<`AxiosResponse`<`AllCastsInThreadResponse`, `any`\>\>
```
Why did all these links in the docs files break?

piratekev replied:
still trying to figure this out

piratekev replied:
the same thing happened in my other PR too

piratekev replied:
When I updated the precommit to run 

```
yarn workspaces foreach --all run generate:openapi-axios
yarn workspaces foreach --all run generate:openapi-typescript
yarn workspaces foreach --all run fixup-imports
yarn workspaces foreach --all run generate:docs
``` 
instead of just
```
yarn workspaces foreach --all run generate
```

that ran `yarn workspaces foreach --all run generate:docs` for the first time in a while for the neyanr workspace. The generate command doesn't invoke the md-magic command currently - see [package.json](https://github.com/standard-crypto/farcaster-js/blob/develop/packages/farcaster-js-neynar/package.json#L86)

```
"generate": "yarn run generate:openapi",
"generate:docs": "md-magic --path './README.md' && typedoc --plugin typedoc-plugin-markdown",
"generate:openapi": "yarn run generate:openapi-axios && yarn run generate:openapi-typescript && yarn run fixup-imports",
```

That said, it still shouldn't just be removing the relative links. The hub-rest workspace still has relative links, so still investigating.

gsgalloway commented on packages/farcaster-js-neynar/README.md:
````diff
@@ -96,7 +96,8 @@ console.log(`Reply hash:${publishedCast.hash}`);
 <!-- The below code snippet is automatically added from ./examples/likeAndRecast.ts -->
 ```ts
-import { NeynarAPIClient, ReactionType } from '@standard-crypto/farcaster-js-neynar';
+import { NeynarAPIClient } from '@standard-crypto/farcaster-js-neynar';
+import { ReactionType } from '@standard-crypto/farcaster-js-neynar/v2';
````
Why did this happen?

piratekev replied:
not sure but fixed

gsgalloway reviewed (changes requested):
Thank you for implementing this! 🙏

piratekev committed:
chore(farcaster-js-hub-rest): refactor & docs

affects: @standard-crypto/farcaster-js-hub-rest

piratekev commented on packages/farcaster-js-hub-rest/tests/integration/hubRestAPIClient.ts:
```diff
@@ -19,10 +19,17 @@ import {
   UserDataType,
 } from '../../src/openapi/index.js';
 import type { paths as SchemaPaths } from '../../src/openapi/generated/schema.js';
+import { hexToBytes } from '@noble/hashes/utils';
```
this line is marked as a relative import, fix added to package.json

gsgalloway replied:
whoa, I'd be curious to know more

piratekev committed:
chore(farcaster-js-hub-rest): review comments

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js-neynar

gsgalloway commented on packages/farcaster-js-hub-rest/src/hubRestApiClient.ts:
```diff
@@ -168,7 +167,7 @@ export class HubRestAPIClient {
     },
     fid: number,
-    signer: NobleEd25519Signer,
+    signer: string,
```
Nice, I agree with your idea to ask the dev for a privkey hex string instead of a typed NobleEd25519Signer object.

I think it will require that we make the parameter name more descriptive now, though, since we lost the documentation that was previously provided by the parameter's type annotation.

What about something like `signerPrivateKeyHex`?

gsgalloway commented on packages/farcaster-js-hub-rest/src/utils.ts:
```diff
@@ -0,0 +1,13 @@
+import { Wallet } from 'ethers';
+
+export function hexToSigner(signerHex: string): NobleEd25519Signer {
+  const privateKeyBytes = hexToBytes(signerHex.slice(2));
```
Thoughts on having this accept inputs both with and without the leading 0x?

gsgalloway commented on examples/hubRest.ts:
```diff
@@ -2,3 +2,10 @@ import { HubRestAPIClient } from '@standard-crypto/farcaster-js';
+const fid = 111;
+const writeClient = new HubRestAPIClient({ hubUrl: 'https://hub.farcaster.standardcrypto.vc:2281' });
+
+const publishCastResponse = await writeClient.submitCast({ text: 'This is a test cast submitted from farcaster-js-hub-rest' }, fid, signerPrivateKey);
```
super nit
```suggestion
const publishCastResponse = await writeClient.submitCast({ text: 'This is a test cast submitted from farcaster-js' }, fid, signerPrivateKey);
```

gsgalloway commented on packages/farcaster-js-hub-rest/package.json:
```diff
@@ -50,7 +50,8 @@
     "cz": "yarn run -T cz",
-    "fixup-imports": "yarn run -T fix-esm-import-path src && sed -i.bak 's|models\"|models/index.js\"|' src/openapi/generated/**/*.ts && sed -i.bak 's|models|models/index.js|' src/openapi/generated/apis/**/*.ts && rm src/openapi/generated/**/*.ts.bak",
+    "fixup-imports": "yarn run -T fix-esm-import-path src && yarn run fix-utils && sed -i.bak 's|models\"|models/index.js\"|' src/openapi/generated/**/*.ts && sed -i.bak 's|models|models/index.js|' src/openapi/generated/apis/**/*.ts && rm src/openapi/generated/**/*.ts.bak",
+    "fix-utils": "sed -i.bak 's/@noble\\/hashes\\/utils\\.js/@noble\\/hashes\\/utils/g' src/utils.ts && rm src/utils.ts.bak",
```
This is curious. `src/utils.ts` was a hand-written file, so why does it need a script to edit one line?

Also, why does that line need to be edited as such?

gsgalloway commented on packages/farcaster-js-hub-rest/src/utils.ts:
```diff
@@ -0,0 +1,13 @@
+  return new NobleEd25519Signer(privateKeyBytes);
+}
+
+export function eip712SignerFromPhrase(mnemonic: string): Eip712Signer {
```
Is this used?

piratekev replied:
It's used in submitVerification to get the address of the account to verify and to sign the verification claim

```
const verificationSigner = eip712SignerFromPhrase(verification.verifiedAddressMnemonic);
const addressBytes = await verificationSigner.getSignerKey();
const ethSignResult = await verificationSigner.signVerificationEthAddressClaim(claim.value);
```

piratekev replied:
https://github.com/standard-crypto/farcaster-js/blob/hub-rest-submit-message/packages/farcaster-js-hub-rest/src/hubRestApiClient.ts#L410

gsgalloway commented on packages/farcaster-js-hub-rest/src/index.ts:
```diff
@@ -2,3 +2,4 @@ export * from './hubRestApiClient.js';
 export * as FarcasterEpochTimestamp from './farcasterEpochTimestamp.js';
 export * from './logger.js';
 export * from './openapi/index.js';
+export * from './utils.js';
```
Do you foresee a developer having use for importing the functions in this file?

piratekev replied:
nope, removed.

piratekev committed:
chore(farcaster-js-hub-rest): review comments

affects: @standard-crypto/farcaster-js-hub-rest

piratekev commented:
@gsgalloway ready for another review here, I'll circle back on the docs generation stuff.

gsgalloway commented on packages/farcaster-js-hub-rest/package.json:
```diff
@@ -12,13 +12,13 @@
   "dependencies": {
-    "@farcaster/core": "^0.13.1"
+    "@farcaster/core": "^0.13.1",
+    "ethers": "^6.8.1"
```
This is a big dependency to introduce. Double-checking that we want/need to. 

I see it used for importing `Wallet` as part of the `eip712SignerFromPhrase` function. Is there another approach that wouldn't pull in ethers? 

I'm cool if the answer is "no," and don't spend a bunch of time on it. Just something to think about, because ethers is both large and finicky if people are already importing a different version of it into their project.

piratekev replied:
updated to replace ethers with viem

piratekev committed:
chore(farcaster-js-hub-rest): review comments

affects: @standard-crypto/farcaster-js-hub-rest

gsgalloway commented on packages/farcaster-js-hub-rest/src/utils.ts:
```diff
@@ -11,9 +11,14 @@ export function hexToSigner(signerHex: string): NobleEd25519Signer {
-  return new EthersEip712Signer(wallet);
+export function eip712SignerFromMnemonicPrPrivateKey(mnemonicPrPrivateKey: string): Eip712Signer {
+  let account;
+  if (mnemonicPrPrivateKey.startsWith('0x')) {
```
Big nitpick here, but the detail that the private key must start with `0x` is a bit hidden from the caller of `submitVerification`.

This probably wouldn't matter so much, except for the fact that some wallets (Metamask for example) will export an account's private key without the leading `0x`

piratekev replied:
updated to work with or without 0x prefix

piratekev committed:
chore(farcaster-js-hub-rest): review comments

affects: @standard-crypto/farcaster-js-hub-rest

piratekev committed:
chore(farcaster-js-hub-rest): add examples, cleanup

affects: @standard-crypto/farcaster-js-hub-rest, @standard-crypto/farcaster-js-neynar,
@standard-crypto/farcaster-js
